rust/protocol/src/storage/traits.rs

```diff
@@ -1,38 +1,70 @@
 //
-// Copyright 2020 Signal Messenger, LLC.
+// Copyright 2020-2021 Signal Messenger, LLC.
 // SPDX-License-Identifier: AGPL-3.0-only
 //
 
+//! Traits defining several stores used throughout the Signal Protocol.
+
+// TODO
+// #![warn(missing_docs)]
+
 use async_trait::async_trait;
 use uuid::Uuid;
 
-use crate::state::{PreKeyId, SignedPreKeyId};
 use crate::{
-    IdentityKey, IdentityKeyPair, PreKeyRecord, ProtocolAddress, Result, SenderKeyRecord,
-    SessionRecord, SignedPreKeyRecord,
+    address::ProtocolAddress,
+    error::Result,
+    sender_keys::SenderKeyRecord,
+    state::{PreKeyId, PreKeyRecord, SessionRecord, SignedPreKeyId, SignedPreKeyRecord},
+    IdentityKey, IdentityKeyPair,
 };
 
+/// Handle to FFI-provided context object.
+///
+/// This object is not manipulated in rust, but is instead passed back to each FFI
+/// method invocation.
 pub type Context = Option<*mut std::ffi::c_void>;
 
+/// Each Signal message can be considered to have exactly two participants, a sender and receiver.
+///
+/// [IdentityKeyStore::is_trusted_identity] uses this to ensure the identity provided is configured
+/// for the appropriate role.
 #[derive(Debug, Clone, Eq, PartialEq)]
 pub enum Direction {
     Sending,
     Receiving,
 }
 
+/// Interface defining the identity store, which may be in-memory, on-disk, etc.
+///
+/// Signal clients usually use the identity store in a [TOFU] manner, but this is not at
+/// all guaranteed.
+///
+/// [TOFU]: https://en.wikipedia.org/wiki/Trust_on_first_use
 #[async_trait(?Send)]
 pub trait IdentityKeyStore {
+    /// Return the single specific identity the store is assumed to represent, with private key.
     async fn get_identity_key_pair(&self, ctx: Context) -> Result<IdentityKeyPair>;
 
+    /// Return a [u32] specific to this store instance.
+    ///
+    /// This local registration id is the same per-device identifier used in [ProtocolAddress] and
+    /// should not change run over run.
     async fn get_local_registration_id(&self, ctx: Context) -> Result<u32>;
 
+    /// Record an identity into the store. The identity is then considered "trusted".
+    ///
+    /// The return value represents whether an existing identity was replaced (`Ok(true)`). If it is
+    /// new or hasn't changed, the return value is `Ok(false)`.
+    // TODO: create a separate enum for this result instead of using a `bool`!
```

Comment:
Yes please. If you don't do this in this PR (and I don't think you should), maybe create an issue for this. I had to search the Java implementation to know what the bool should be :-)

Reply:
I have modified this TODO to point to the issue #295 I just made for this! :D Thank you!